### PR TITLE
Adding support for linting QcStatements

### PR DESCRIPTION
**Pull request by mtgag**
EU qualified certificates can be used for web site authentication. They use the qcstatements certificate extension. The corresponding specification is ETSI 319 412-5 V2.2.1.

https://www.etsi.org/deliver/etsi_en/319400_319499/31941205/02.02.01_60/en_31941205v020201p.pdf

This fork implements lint checks for this extension according to the above specification.

### CODE REVIEW

File: util/qc_stmt.go [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+		{
+			return "IdEtsiQcsQcEuPDS"
+		}
+	case oi.Equal(IdEtsiQcsQcType):
+		{
+			return "IdEtsiQcsQcType"
+		}
+	default:
+		{
+			panic("unsresolved ETSI QC Statement OID")
```
Comment (justinbastress) [inline util/qc_stmt.go]: Typo
```suggestion
			panic("unresolved ETSI QC Statement OID")
```

File: lints/lint_qcstatem_etsi_type_as_statem_test.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import (
+	"testing"
+)
+
+func TestEtsiTypeAsQcStmt(t *testing.T) {
```
Comment (justinbastress) [inline lints/lint_qcstatem_etsi_type_as_statem_test.go]: It seems like pretty consistently `Statem` is used as the abbreviation for `Statement`, but here it has `Stmt`?

`Statem` presents its own problems (e.g. concating `Statem` and `etsi` looks like a typo for `Statements` with the `n` missing), but it seems to win out over `Stmt` in this PR)

File: lints/lint_qcstatem_qctype_web.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		}
+	} else {
+		return &LintResult{Status: Error, Details: errString}
+	}
+}
+
+func init() {
+	RegisterLint(&Lint{
+		Name:          "w_qcstatem_qctype_web",
+		Description:   "Checks that a QC Statement of the type Id-etsi-qcs-QcType features features at least the type ",
```
Comment (justinbastress) [inline lints/lint_qcstatem_qctype_web.go]:
```suggestion
		Description:   "Checks that a QC Statement of the type Id-etsi-qcs-QcType features features at least the type IdEtsiQcsQctWeb",
```

File: lints/lint_qcstatem_qctype_valid.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	errString += s.GetErrorInfo()
+	if len(errString) == 0 {
+		qcType := s.(util.Etsi423QcType)
+		if len(qcType.TypeOids) == 0 {
+			errString += "no QcType present, sequence of OIDs is empty"
+		}
+		for _, t := range qcType.TypeOids {
+
+			if !t.Equal(util.IdEtsiQcsQctEsign) && !t.Equal(util.IdEtsiQcsQctEseal) && !t.Equal(util.IdEtsiQcsQctWeb) {
+				if len(errString) > 0 {
```
Comment (justinbastress) [inline lints/lint_qcstatem_qctype_valid.go]:
```diff
-				if len(errString) > 0 {
-  					errString += "; "
- 				}
-				errString += fmt.Sprintf("encountered invalid ETSI QcType OID: %v", t)
+				util.AppendToStringSemicolonDelim(&errString, fmt.Sprintf("encountered invalid ETSI QcType OID: %v", t))
```

**Comment by justinbastress** [comment]
I don't see any substantial issues here -- beyond the style comments about, there were a few other non-standard abbreviations that weren't clear until I looked around for context (like `Oi` for `OBJECT IDENTIFIER`, rather than the more common `Oid`, or `Wo` for "without"), but that's more a matter of preference.

File: lints/lint_qcstatem_etsi_type_as_statem.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+package lints
+
+import (
+	"encoding/asn1"
+	"fmt"
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/util"
+)
+
+type qcStatemetsiTypeAsStatem struct{}
```
[the remark this answers is not in the document]
Comment (mtgag) [reply lints/lint_qcstatem_etsi_type_as_statem.go]: Yes, you are right.

File: lints/lint_qcstatem_qctype_valid.go (hunk 4 above)
Comment (mtgag) [reply lints/lint_qcstatem_qctype_valid.go]: yes. This change is better.

File: lints/lint_qcstatem_etsi_type_as_statem_test.go (hunk 2 above)
Comment (mtgag) [reply lints/lint_qcstatem_etsi_type_as_statem_test.go]: Yes, this should be changed to Statem

**Comment by mtgag** [comment]
We resolved the copyright issue by deleting the corresponding line.

**Pull request merged**